File: 240618/정수 n개의 합 3/sum-of-n-integers-3.py
import sys

n, k = tuple(map(int, input().split()))

grid = [[0] * (n + 1)] + [
    [0] + list(map(int, input().split()))
    for _ in range(n)
]

prefix_sum = [
    [0] * (n + 1)
    for _ in range(n + 1)
]

answer = -sys.maxsize

for i in range(1, n + 1):
    for j in range(1, n + 1):
        prefix_sum[i][j] = prefix_sum[i-1][j] + prefix_sum[i][j-1] - prefix_sum[i-1][j-1] + grid[i][j]

# k*k 사각형 중 최대합 찾기
for i in range(k, n + 1):
    for j in range(k, n + 1):
        temp = prefix_sum[i][j] - prefix_sum[i-k][j] - prefix_sum[i][j-k] + prefix_sum[i-k][j-k]
        answer = max(answer, temp)

print(answer)

File: 250730/정수 N개의 합 3/sum-of-n-integers-3.js
const fs = require("fs");
const input = fs.readFileSync(0).toString().trim().split('\n');

const [n, k] = input[0].trim().split(' ').map(Number);
const grid = input.slice(1, 1 + n).map(line => line.trim().split(' ').map(Number));

// Please Write your code here.
const prefixSum = Array.from({ length: n + 1 }, () => Array(n + 1).fill(0));
const maxGrid = Array.from({ length: n + 1 }, () => Array(n + 1).fill(0));
let ans = 0;

// 초기화
prefixSum[1][1] = grid[0][0]

for (let i = 1; i <= n; i++) {
    for (let j = 1; j <= n; j++) {
        prefixSum[i][j] = prefixSum[i - 1][j] + prefixSum[i][j - 1] - prefixSum[i - 1][j - 1] + grid[i - 1][j - 1];
    }
}

for (let i = k; i <= n; i++) {
    for (let j = k; j <= n; j++) {
        maxGrid[i][j] = prefixSum[i][j] - prefixSum[i - k][j] - prefixSum[i][j - k] + prefixSum[i - k][j - k] 
    }
}

for (let i = 0; i <= n; i++) {
    for (let j = 0; j <= n; j++) {
        ans = Math.max(ans, maxGrid[i][j])
    }
}

console.log(ans);